File: python/tests/test_metrics.py
import os
import tempfile
import pytest
from scanner.metrics import analyze_file


def make_file(content, suffix='.py'):
    f = tempfile.NamedTemporaryFile(mode='w', suffix=suffix, delete=False,
                                    encoding='utf-8')
    f.write(content)
    f.close()
    return f.name


def test_python_counts():
    path = make_file('# comment\ndef foo():\n    pass\n\nx = 1\n')
    try:
        r = analyze_file(path, 'Python')
        assert r['comment_lines'] == 1
        assert r['loc'] == 3     # def foo():, pass, x = 1
        assert r['blank_lines'] == 1
        assert r['functions'] == 1
        assert r['classes'] == 0
    finally:
        os.unlink(path)


def test_blank_lines():
    path = make_file('\n\nx = 1\n\n')
    try:
        r = analyze_file(path, 'Python')
        assert r['blank_lines'] == 3
        assert r['loc'] == 1
    finally:
        os.unlink(path)


def test_js_comment_lines():
    path = make_file('// comment\nfunction foo() {\n    return 1;\n}\n', '.js')
    try:
        r = analyze_file(path, 'JavaScript')
        assert r['comment_lines'] == 1
        assert r['functions'] >= 1
    finally:
        os.unlink(path)


def test_inline_comment_counts_as_loc():
    path = make_file('x = 1  # inline comment\n')
    try:
        r = analyze_file(path, 'Python')
        assert r['loc'] == 1
        assert r['comment_lines'] == 0
    finally:
        os.unlink(path)


def test_class_count():
    path = make_file('class Foo:\n    pass\n\nclass Bar:\n    pass\n')
    try:
        r = analyze_file(path, 'Python')
        assert r['classes'] == 2
    finally:
        os.unlink(path)


def test_python_block_comment():
    path = make_file('def foo():\n    """\n    docstring\n    """\n    pass\n')
    try:
        r = analyze_file(path, 'Python')
        # The """ lines count as comment_lines, not loc
        assert r['functions'] == 1
        assert r['comment_lines'] >= 1
    finally:
        os.unlink(path)


def test_js_block_comment():
    content = '/* block\n   comment */\nconst x = 1;\n'
    path = make_file(content, '.js')
    try:
        r = analyze_file(path, 'JavaScript')
        assert r['comment_lines'] >= 1
        assert r['loc'] >= 1
    finally:
        os.unlink(path)


def test_unreadable_returns_none():
    r = analyze_file('/nonexistent/path/file.py', 'Python')
    assert r is None


def test_function_and_class():
    content = 'class Foo:\n    def bar(self):\n        pass\n'
    path = make_file(content)
    try:
        r = analyze_file(path, 'Python')
        assert r['functions'] == 1
        assert r['classes'] == 1
    finally:
        os.unlink(path)

File: javascript/tests/test_metrics.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { mkdtempSync, writeFileSync, unlinkSync } from 'node:fs';
import { join } from 'node:path';
import { tmpdir } from 'node:os';
import { analyzeFile, countLines } from '../src/metrics.js';

function makeFile(content, ext = '.py') {
  const dir = mkdtempSync(join(tmpdir(), 'scanner-test-'));
  const path = join(dir, `file${ext}`);
  writeFileSync(path, content, 'utf8');
  return path;
}

test('python comment and code lines', () => {
  const path = makeFile('# comment\ndef foo():\n    pass\n\nx = 1\n');
  const r = analyzeFile(path, 'Python');
  assert.equal(r.comment_lines, 1);
  assert.equal(r.loc, 3);
  assert.equal(r.blank_lines, 1);
  assert.equal(r.functions, 1);
});

test('blank lines', () => {
  const path = makeFile('\n\nx = 1\n\n');
  const r = analyzeFile(path, 'Python');
  assert.ok(r.blank_lines >= 3);
  assert.equal(r.loc, 1);
});

test('js comment lines', () => {
  const path = makeFile('// comment\nfunction foo() {\n    return 1;\n}\n', '.js');
  const r = analyzeFile(path, 'JavaScript');
  assert.equal(r.comment_lines, 1);
  assert.ok(r.functions >= 1);
});

test('inline comment counts as loc', () => {
  const path = makeFile('x = 1  # inline comment\n');
  const r = analyzeFile(path, 'Python');
  assert.equal(r.loc, 1);
  assert.equal(r.comment_lines, 0);
});

test('class count', () => {
  const path = makeFile('class Foo:\n    pass\n\nclass Bar:\n    pass\n');
  const r = analyzeFile(path, 'Python');
  assert.equal(r.classes, 2);
});

test('unreadable file returns null', () => {
  const r = analyzeFile('/nonexistent/path/file.py', 'Python');
  assert.equal(r, null);
});

test('function and class together', () => {
  const path = makeFile('class Foo:\n    def bar(self):\n        pass\n');
  const r = analyzeFile(path, 'Python');
  assert.equal(r.functions, 1);
  assert.equal(r.classes, 1);
});

test('js block comment', () => {
  const path = makeFile('/* block\n   comment */\nconst x = 1;\n', '.js');
  const r = analyzeFile(path, 'JavaScript');
  assert.ok(r.comment_lines >= 1);
  assert.ok(r.loc >= 1);
});
